Add: openPMD-viewer

ax3l opened the pull request:
Add the `py-openpmd-viewer` package.

ax3l committed:
Add: openPMD-viewer

Add the `py-openpmd-viewer` package.

adamjstewart commented on var/spack/repos/builtin/packages/py-openpmd-viewer/package.py:
```diff
@@ -0,0 +1,44 @@
+    """Python visualization tools for openPMD files"""
+
+    homepage = "https://www.openPMD.org"
+    url      = "https://github.com/openPMD/openPMD-viewer/archive/refs/tags/1.2.0.tar.gz"
```
https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

ax3l replied:
Thanks! Since @RemiLehe and I maintain this, mainly point 1 applies to us. Can update to use our pip releases nonetheless for consistency :)

adamjstewart commented on var/spack/repos/builtin/packages/py-openpmd-viewer/package.py:
```diff
@@ -0,0 +1,44 @@
+    depends_on('python@3.6:', type=('build', 'run'))
+    depends_on('py-h5py@2.8.0:', type=('build', 'run'))
+    depends_on('py-matplotlib', type=('build', 'run'))
+    depends_on('py-numpy@1.15:', type=('build', 'run'))
```
```suggestion
    depends_on('py-numpy@1.15:1', type=('build', 'run'))
```

ax3l replied:
You probably mean this :)
```suggestion
    depends_on('py-numpy@1.15:2', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-openpmd-viewer/package.py:
```diff
@@ -0,0 +1,44 @@
+
+    depends_on('python@3.6:', type=('build', 'run'))
+    depends_on('py-h5py@2.8.0:', type=('build', 'run'))
+    depends_on('py-matplotlib', type=('build', 'run'))
```
matplotlib appears to be optional, I would remove it or add a "plot" variant.

ax3l replied:
@RemiLehe what do you think? I would tend to at most make this a `plot` variant and make it default-on in Spack.

adamjstewart commented on var/spack/repos/builtin/packages/py-openpmd-viewer/package.py:
```diff
@@ -0,0 +1,44 @@
+    depends_on('py-scipy', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
+    depends_on('py-tqdm', type=('build', 'run'))
+    depends_on('py-wget', type=('build', 'run'))  # Tutorials
```
wget appears to be optional, I would remove it or add a "tutorials" variant.

adamjstewart commented on var/spack/repos/builtin/packages/py-openpmd-viewer/package.py:
```diff
@@ -0,0 +1,44 @@
+
+    version('1.2.0', sha256='a27f8ac522c4c76fd774095e156a8b280c9211128f50aa07f16ac70d8222384d')
+
+    variant('jupyter', default=False,
+            description='Enable Jupyter Widget GUI')
+    variant('numba', default=False,
+            description='Enable accelerated depositions for histograms')
```
Usually variants in Python packages match "extras_require" in the `setup.py`, but these seem to be custom.

ax3l replied:
Yes, forgotten in the repo. Fix in: https://github.com/openPMD/openPMD-viewer/pull/327

ax3l committed:
openPMD-viewer: Improve Variant Control

ax3l committed:
openPMD-viewer: Update to Source from Pip/PyPA

https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

ax3l commented on var/spack/repos/builtin/packages/py-openpmd-viewer/package.py:
```diff
@@ -0,0 +1,57 @@
+    with when('+plot'):
+        depends_on('py-matplotlib', type=('build', 'run'))
+        # missing in Spack:
+        #with when('+jupyter'):
+        #    depends_on('py-ipympl', type=('build', 'run'))
```
```suggestion
        # with when('+jupyter'):
        #     depends_on('py-ipympl', type=('build', 'run'))
```

ax3l committed:
Fix style: comment

adamjstewart commented on var/spack/repos/builtin/packages/py-openpmd-viewer/package.py:
```diff
@@ -0,0 +1,57 @@
+    variant('tutorials', default=True,
+            description='Enable dependencies for tutorials')
+
+    depends_on('python@3.6:', type=('build', 'run'))
```
```suggestion
    depends_on('python@3.5:', type=('build', 'run'))
```
According to the `setup.py`

adamjstewart commented on var/spack/repos/builtin/packages/py-openpmd-viewer/package.py:
```diff
@@ -0,0 +1,57 @@
+    depends_on('py-tqdm', type=('build', 'run'))
+
+    with when('backend=h5py'):
+        depends_on('py-h5py@2.8.0:', type=('build', 'run'))
```
According to the `setup.py` this is currently required. If it isn't installed, setuptools/pip will install it for you, which won't work on air-gapped systems.

adamjstewart reviewed (commented):
I think a few things in this recipe are designed for develop, or designed for what the `setup.py` _should_ look like as opposed to what it actually looks like.

ax3l committed:
openPMD-viewer: fix requirements

add reviewer feedback

ax3l commented:
Thank you for the review @adamjstewart, much appreciated :+1:
